libraries/Wire/src/Wire.cpp

```diff
@@ -132,6 +135,81 @@ bool TwoWire::setPins(int sdaPin, int sclPin)
     return !i2cIsInit(num);
 }
 
+bool TwoWire::allocateWireBuffer(void)
+{
+    // or both buffer can be allocated or none will be
+    if (rxBuffer == NULL) {
+            rxBuffer = (uint8_t *)malloc(BufferSize);
+            if (rxBuffer == NULL) {
+                log_e("Can't allocate memory for I2C_%d rxBuffer", num);
+                return false;
+            }
+    }
+    if (txBuffer == NULL) {
+            txBuffer = (uint8_t *)malloc(BufferSize);
+            if (txBuffer == NULL) {
+                log_e("Can't allocate memory for I2C_%d txBuffer", num);
+                freeWireBuffer();  // free rxBuffer for safety!
+                return false;
+            }
+    }
+    // in case both were allocated before, they must have the same size. All good.
+    return true;
+}
+
+void TwoWire::freeWireBuffer(void)
+{
+    if (rxBuffer != NULL) {
+        free(rxBuffer);
+        rxBuffer = NULL;
+    }
+    if (txBuffer != NULL) {
+        free(txBuffer);
+        txBuffer = NULL;
+    }
+}
+
+size_t TwoWire::setBufferSize(size_t bSize)
+{
+    // Maximum size .... HEAP limited ;-)
+    if (bSize < 32) {    // 32 bytes is the I2C FIFO Len for ESP32/S2/S3/C3
+        log_e("Minimum Wire Buffer size is 32 bytes");
+        return 0;
+    }
+
+#if !CONFIG_DISABLE_HAL_LOCKS
+    if(lock != NULL){
+        //acquire lock
+        if(xSemaphoreTake(lock, portMAX_DELAY) != pdTRUE){
+            log_e("could not acquire lock");
+            return 0;
+        }
+#endif
+    // allocateWireBuffer allocates memory for both pointers or just free them
+    if (rxBuffer != NULL || txBuffer != NULL) {
+        // if begin() has been already executed, memory size changes... data may be lost. We don't care! :^)
+        if (bSize != BufferSize) {
+            // we want a new buffer size ... just reset buffer pointers and allocate new ones
+            freeWireBuffer();
+            BufferSize = bSize;
+            if (!allocateWireBuffer()) {
+                // failed! Error message already issued
+                bSize = 0; // returns error
+            }
+        } // else nothing changes, all set!
+    } else {
+        // no memory allocated yet, just change the size value - allocation in begin()
+        BufferSize = bSize;
+    }
+#if !CONFIG_DISABLE_HAL_LOCKS
+        //release lock
+        xSemaphoreGive(lock);
+    }
+#endif
+    return bSize;
```

Comment:
It makes more sense for the return to be `BufferSize` to be able to check, if the function made the change :)

Reply:
In case it fails by `malloc()` problem, `bufferSize` will keep the requested size and `bSize` will go to zero, as return value. So, with a `begin()` it will still be able to try to allocate the memory based on `bufferSize`.

Reply:
Lines 207 ... 210 shall solve the issue of `setBufferSize()` called before `begin()`

Reply:
Ok I got it, we can leave the `bSize` now :) It was not making sense before the lock change. It was returning inputed `bSize` even when `bufferSize` did not changed :) So marking as solved